Add readable id in searchable fields of wagtail

arslanashraf7 opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
- [x] Migrations (Wagtail Search App's)
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
#351 

**Additional Note - Reviewer Note**
- While working on this, I also noticed that instead of showing the title from `CoursePage` model of wagtail, we were showing the title of the `Course` model from `courses` app. So I fixed that as well. This means updating the title of a course in `/cms/pages` should be displayed everywhere instead of the title from the Course model(Can be changed in Django Admin).
- The wagtail index search doesn't update the index of the related keys automatically, so in the case when you update the readable id of a course manually from admin after the course has been added in CMS, you'll need to run `./manage.py wagtail_update_index` so that the related `readable_id` index updates

#### What's this PR do?
- Adds the readable id to be part of search in wagtail

#### How should this be manually tested?
- Search any course with the readable id instead of title and you should see the matching course in the list
- Search a course by title, it should be returned in the search results

#### Any background context you want to provide?
- Readable id was added in `admin_display_title` in one of recent PR(https://github.com/mitodl/mitxonline/pull/336), where we formatted the course pages display title in addition with readable id. 

#### Screenshots (if appropriate)
<img width="1333" alt="Screenshot 2022-01-04 at 2 04 46 PM" src="https://user-images.githubusercontent.com/34372316/148035842-a82c3c90-887a-4a3a-aa0e-a43b20d1f2d7.png">
<img width="1277" alt="Screenshot 2022-01-04 at 2 05 03 PM" src="https://user-images.githubusercontent.com/34372316/148035847-a7870a29-c76e-4306-90fa-42220088a5aa.png">

arslanashraf7 commented:
@briangrossman Could you please take a look at the screenshots attached to the PR description along with `Reviewer Note` that i added for details that you might want to take a look at.

Extending to the Reviewer note above, we will need to look for the following as well:
-  We need to run `./manage.py wagtail_update_index` just once when this build is deployed (To create the initial index for `readable_id` of the existing courses)
- The index for `readable_id` would update automatically when we create a new course page in CMS.
- We might need to run a cron job occasionally to update the index **_But_** that might not really be necessary as I think the frequency of changing the `readable_id` is quite low and even if we need to change it after the page has been created in CMS, we can just run `./manage.py wagtail_update_index` and the index should rebuild. (Only applies if a course was created in CMS already and its readable id was changed in Django Admin later on.) So I think we can run this manually when needed.


@briangrossman let me know what you think about the above points.

briangrossman commented:
@arslanashraf7 
The screenshots look good. Thanks for taking care of this--it looks like it was not trivial.

We'll have to remember the notes above for when to run `wagtail_update_index`

arslanashraf7 commented:
>The screenshots look good. Thanks for taking care of this--it looks like it was not trivial.
@briangrossman 

Yes, I reverted my last workaround of using signals and stuff for automatically running the command and instead used the simpler implementation where we can run the command manually. Which made sense based on the low probability of `readable_id` getting changed.

And I hope you had a look at below(Is my below assumption correct?):
>While working on this, I also noticed that instead of showing the title from CoursePage model of wagtail, we were showing the title of the Course model from courses app. So I fixed that as well. This means updating the title of a course in /cms/pages should be displayed everywhere instead of the title from the Course model(Can be changed in Django Admin).

briangrossman commented:
@arslanashraf7 
> And I hope you had a look at below(Is my below assumption correct?):
> 
> > While working on this, I also noticed that instead of showing the title from CoursePage model of wagtail, we were showing the title of the Course model from courses app. So I fixed that as well. This means updating the title of a course in /cms/pages should be displayed everywhere instead of the title from the Course model(Can be changed in Django Admin).

Yes. This definitely seems like the right implementation. Thanks for that.

briangrossman commented:
I reviewed this on RC and it LGTM.

Note that @arslanashraf7  and I had some discussion about the nuances of searching. Just documenting them here:

**Question**
Here’s a couple examples that surprised me:
- Searching for `course` returns all courses with the `readable_id` that begins with `course-v1`. Searching for `cours` returns nothing. I assume this has something to do with the way the search indexes.
- Searching for `+14.750x` returns the course with the `readable_id` `course-v1:MITx+14.750x`, but searching for `14.750x` returns nothing

**Answer**
[...] for the cases you mentioned i think there are a few limitations mentioned for the partial search on the site [documentation](https://docs.wagtail.io/en/v2.13.4/reference/contrib/postgres_search.html#postgres-search). so, what it is doing is because of the how [boost/postgres search](https://www.postgresql.org/docs/current/textsearch.html) is working underneath.

arslanashraf7 commented:
Thanks for adding that comment @briangrossman.

_Just for future references_, I'd add just one additional thing related to when we say `Partial Matching` isn't supported & but `wagtail supports all the text search features of Postgres` as mentioned in [answer here](https://github.com/mitodl/mitxonline/pull/352#issuecomment-1006813221). More clarity on the use-cases below:

>Searching for course returns all courses with the readable_id that begins with course-v1. Searching for cours returns nothing. I assume this has something to do with the way the search indexes.

It still searches the whole text within the fields `title & readable_id`, and the fact that it returns courses on `course-v1` is related to colon (`:` or `<space>`) somehow(It divides the words and index them for search), so whenever there is a colon within the text it divides the text into two and able to search both parts independently that's why we see courses on `course-v1`. (I looked around to find the docs on this but i couldn't but I understood the behavior)

>Searching for +14.750x returns the course with the readable_id course-v1:MITx+14.750x, but searching for 14.750x returns nothing

Same case as mentioned above.
